src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1670,7 +1663,16 @@ public SolutionCompilationState WithDocumentText(IEnumerable<DocumentId?> docume
 
             var documentState = this.SolutionState.GetProjectState(documentId.ProjectId)?.DocumentStates.GetState(documentId);
             if (documentState != null)
-                result = result.WithDocumentText(documentId, text, mode);
+            {
+                // before allocating an array below (and calling into a function that does a fair amount of linq work),
+                // do a fast check if the text has actually changed. this shows up in allocation traces and is
+                // worthwhile to avoid for the common case where we're continually being asked to update the same doc to
+                // the same text (for example, when GetOpenDocumentInCurrentContextWithChanges) is called.
+                //
+                // The use of GetRequiredState mirrors what happens in WithDocumentTexts
+                if (!SourceTextIsUnchanged(documentState, text))
```

Comment:
>if (!SourceTextIsUnchanged(documentState, text)) [](http://example.com/codeflow?start=16&length=48)

Sorry, another question. 

Is there usually only a single documentId that will end up calling WithDocumentTexts? (if multiple will, it seems like they should be batched together)

Reply:
Yes.  That definitely makes sense. I can tweak to that when I get home.  Lmk if this is otherwise ok

Reply:
done.